premailer/cache.py

```diff
@@ -1,83 +1,61 @@
-        the argument to the wrapped function must be hashable else
+        function_cache is a decorator for caching function call.
+        The argument to the wrapped function must be hashable else
         it will not work
 
-        expected_max_entries is for protecting cache failure. If cache
-        misses more than this number the cache will turn off itself.
-        Specify None you sure that the cache  will not cause memory
-        limit problem.
-
-        Args:
-            expected_max_entries(integer OR None): will raise if not correct
+        Function wrapper accepts max_cache_entries argument which specifies
+        the maximum number of different data to cache. Specifying None will not
+        limit the size of the cache at all.
 
         Returns:
             function
 
     """
-    if (
-        expected_max_entries is not None and
-        not isinstance(expected_max_entries, int)
-    ):
-        raise TypeError(
-            'Expected expected_max_entries to be an integer or None'
-        )
 
     # indicator of cache missed
     sentinel = object()
 
     def decorator(func):
-        cached = _Cache()
+        cache = {}
 
         @functools.wraps(func)
         def inner(*args, **kwargs):
-            if cached.off:
+            max_cache_entries = kwargs.pop('max_cache_entries', 1000)
+
+            if (
+                max_cache_entries is not None and
+                not isinstance(max_cache_entries, int)
+            ):
+                raise TypeError(
+                    'Expected max_cache_entries to be an integer or None'
+                )
+
+            if max_cache_entries == 0:
                 return func(*args, **kwargs)
 
-            keys = args
+            keys = []
+            for arg in args:
+                if isinstance(arg, list):
+                    keys.append(tuple(arg))
+                else:
+                    keys.append(arg)
+
             if kwargs:
                 sorted_items = sorted(kwargs.items())
                 for item in sorted_items:
-                    keys += item
+                    keys.append(item)
 
-            hashed = hash(_HashedSeq(keys))
-            result = cached.cache.get(hashed, sentinel)
+            hashed = hash(tuple(keys))
+            result = cache.get(hashed, sentinel)
             if result is sentinel:
-                cached.missed += 1
                 result = func(*args, **kwargs)
-                cached.cache[hashed] = result
-                # # something is wrong if we are here more than expected
-                # # empty and turn it off
-                if (
-                    expected_max_entries is not None and
-                    cached.missed > expected_max_entries
-                ):
-                    cached.off = True
-                    cached.cache.clear()
+
+                if max_cache_entries is None or len(cache) < max_cache_entries:
+                    cache[hashed] = result
```

Comment:
It's a bit strange that the caching just decides to stubbornly refuse to add more entries. I've already forgotten what you said in the justification in the PR description. Can you please add a code comment here to explain this reasoning.